### PR TITLE
[web][api][SIMS #1018]API Reorganization for Manage Users - Part 3

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Institution user-related endpoints moved to a new controller institution-user.
- Endpoint no longer using `:userName` and using `:institutionUserId` and proposed on the previous PR.
- Ministry users are now able to disable/edit users in a way that the admin no admin will be available.
- Update audit fields for when a new institution is created by the institution used for the first time, when the permissions are updated, and when new users are created using the Manage users.
- The institution login screen was updated to remove the "business" word and also adjusted to be more close to Figma.

image

- Fixed the user synchronization from BCeID that ensures that the read-only fields below are always updated upon user login.

image

- Fixed the institution synchronization from BCeID that ensures that the read-only fields below are always updated upon user login.

image

### CODE REVIEW

File: sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts [hunk 1]
```diff
@@ -217,4 +217,33 @@ export class InstitutionLocationInstitutionsController extends BaseController {
...
+        userToken.authorizations.getLocationsIds(),
+      );
+    return InstitutionLocationData.map((location) => {
+      return {
+        id: location.id,
+        name: location.name,
+        address: {
+          addressLine1: location.data.address?.addressLine1,
+          addressLine2: location.data.address?.addressLine2,
+          provinceState: location.data.address?.provinceState,
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts:241]: just a qn, we do have a different story to update to the new address format in other forms right?

File: sources/packages/api/src/route-controllers/institution-user/institution-user.controller.service.ts [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+    paginationOptions: PaginationOptions,
+  ): Promise<PaginatedResults<InstitutionUserAPIOutDTO>> {
+    const [institutionUsers, count] =
+      await this.institutionService.getInstitutionUsers(
+        institutionId,
+        paginationOptions,
+      );
+
+    return {
+      results: institutionUsers.map((institutionUser: InstitutionUser) => {
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/institution-user/institution-user.controller.service.ts:66]: to simply ` results: institutionUsers.map((institutionUser: InstitutionUser) => 
        this.transformToInstitutionUserAPIOutDTO(institutionUser)
      }),`

File: sources/packages/api/src/route-controllers/institution-user/institution-user.institutions.controller.ts [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+  @ApiNotFoundResponse({
+    description: "User not found.",
+  })
+  @ApiForbiddenResponse({
+    description:
+      "Details requested for user who does not belong to the institution.",
+  })
+  @IsInstitutionAdmin()
+  @Get(":institutionUserId")
+  async getInstitutionUserByUserName(
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/institution-user/institution-user.institutions.controller.ts:240]: you missed to rename the function 😉

File: sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts:241]: I am not sure. Maybe during a UI review but I do not believe that we have something specific.

File: sources/packages/api/src/route-controllers/institution-user/institution-user.institutions.controller.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/institution-user/institution-user.institutions.controller.ts:240]: Good catch, changing it 👍

Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/institution-user/institution-user.institutions.controller.ts]: Renamed!

File: sources/packages/api/src/route-controllers/institution-user/institution-user.controller.service.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/institution-user/institution-user.controller.service.ts]: Changed!

File: sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts:241]: Apparently, this address is returned only to be saved to the Vue store but from there is not consumed anywhere. I would be considering removing it. Thoughts?

File: sources/packages/api/src/services/institution/institution.service.ts [hunk 4]
```diff
@@ -199,12 +167,13 @@ export class InstitutionService extends RecordDataModelService<Institution> {
     institutionModel: InstitutionFormModel,
     userInfo: UserInfo,
   ): Promise<Institution> {
+    // New user to be associated with the institution. It will also be considered the audit user.
```
Comment (ann-aot) [inline sources/packages/api/src/services/institution/institution.service.ts:170]: i didn't get this point `It will also be considered the audit user`

File: sources/packages/api/src/services/institution/institution.service.ts [hunk 5]
```diff
@@ -447,27 +444,37 @@ export class InstitutionService extends RecordDataModelService<Institution> {
...
-    const validateAtLeastOneAdmin = this.validateAtLeastOneAdmin(
-      institutionId,
-      institutionUserId,
-      permissions,
-    );
-    const validateUniqueSigningAuthority = this.validateUniqueSigningAuthority(
-      institutionId,
-      permissions,
-      institutionUserId,
+    const validations: Promise<void>[] = [];
```
Comment (ann-aot) [inline sources/packages/api/src/services/institution/institution.service.ts:458]: for my knowledge, what is the benefit of `:Promise<void>[] = []` with `:[] = []`

File: sources/packages/api/src/services/institution/institution.service.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/institution/institution.service.ts:170]: This is the process that happens when a new BCeID business user is also creating the institution.
Basically at this point, the `userToken` does not have the `userId` because the user is not on the DB yet.
So, the user being created here will be the user also to be used as the audit user. Does it sound better now?

File: sources/packages/api/src/services/institution/institution.service.ts (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/institution/institution.service.ts:458]: If I understood the question, `:[] = []` will be an array of `any` and `Promise<void>[]` will be a typed array of `:Promise<void>`.

File: sources/packages/web/src/services/InstitutionUserService.ts [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    const userType = isAdmin
+      ? InstitutionUserTypes.admin
+      : InstitutionUserTypes.user;
+    const locations: string[] = [];
+    if (isAdmin) {
+      locations.push("All");
+    } else {
+      const locationNames = institutionUser.authorizations
+        .filter((authorization) => !!authorization.location)
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (ann-aot) [inline sources/packages/web/src/services/InstitutionUserService.ts:61]: do we need to disable this? can't we use `authorization.location?.name`

File: sources/packages/web/src/services/InstitutionUserService.ts [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+   * @param isLegalSigningAuthority for admin users, define if the role
+   * legal-signing-authority should be associated with the user.
+   * @param locationAuthorizations for non-admin users, the individual permission
+   * for every location.
+   */
+  private createUserPermissions(
+    isAdmin: boolean,
+    isLegalSigningAuthority: boolean,
+    locationAuthorizations: LocationAuthorization[],
+  ): UserPermissionAPIInDTO[] {
```
Comment (ann-aot) [inline sources/packages/web/src/services/InstitutionUserService.ts:116]: return comment?

File: sources/packages/web/src/services/InstitutionUserService.ts [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+        .filter(
+          (locationAccess) =>
+            locationAccess.userAccess === LocationUserAccess.User,
+        )
+        .map(
+          (locationAccess) =>
+            ({
+              locationId: locationAccess.id,
+              userType: locationAccess.userAccess,
+            } as UserPermissionAPIInDTO),
```
Comment (ann-aot) [inline sources/packages/web/src/services/InstitutionUserService.ts:142]: not part of this PR, but we can remove `as UserPermissionAPIInDTO`

File: sources/packages/web/src/services/InstitutionUserService.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/InstitutionUserService.ts:61]: Without the "!" it is not able to recognize that the '.filter. is already ensuring that all the locations will have a name.
image

File: sources/packages/web/src/services/InstitutionUserService.ts [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+      institutionUserId,
+    );
+  }
+
+  /**
+   * Update the active status of the user.
+   * @param institutionUserId institution user id to have the permissions updated.
+   * @param isActive enable or disable the user.
+   */
+  async updateUserStatus(institutionUserId: number, isActive: boolean) {
```
Comment (ann-aot) [inline sources/packages/web/src/services/InstitutionUserService.ts:225]: not part of this PR, but its missing `: Promise<void> {`

File: sources/packages/web/src/services/http/InstitutionLocation.ts [hunk 10]
```diff
@@ -92,4 +57,17 @@ export class InstitutionLocationApi extends HttpBaseClient {
       ),
     );
   }
+
+  /**
+   * Get location details of logged in user.
+   * @returns location details.
+   */
+  async getMyInstitutionLocationsDetails(
+    header?: any,
```
Comment (ann-aot) [inline sources/packages/web/src/services/http/InstitutionLocation.ts:66]: since you commented this function. can we add `header in the comment

File: sources/packages/web/src/services/http/InstitutionUserApi.ts [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+import { AuthService } from "../AuthService";
+import { getPaginationQueryString } from "@/helpers";
+
+export class InstitutionUserApi extends HttpBaseClient {
+  /**
+   * Get details of user who is logged in.
+   * @returns user details.
+   */
+  async getMyInstitutionDetails(
+    header?: any,
```
Comment (ann-aot) [inline sources/packages/web/src/services/http/InstitutionUserApi.ts:20]: param comment

File: sources/packages/web/src/services/InstitutionUserService.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/InstitutionUserService.ts:142]: It can be removed, but in the cases outside the controller DTOs, like this one, IMO it makes it easier to read.
Open for thoughts. 😉

File: sources/packages/web/src/services/http/InstitutionUserApi.ts [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+      this.addClientRoot(url),
+    );
+  }
+
+  /**
+   * Create a user, associate with the institution, and assign the authorizations.
+   * @param payload authorizations to be associated with the user.
+   * @param institutionId institution to have the user associated. If not provided the
+   * token information will be used, if available.
+   * @returns Primary identifier of the created resource.
```
Comment (ann-aot) [inline sources/packages/web/src/services/http/InstitutionUserApi.ts:55]: there is no return right?

File: sources/packages/web/src/services/InstitutionUserService.ts (hunk 9 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/InstitutionUserService.ts:225]: I will add it.

File: sources/packages/web/src/services/http/InstitutionLocation.ts (hunk 10 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/InstitutionLocation.ts:66]: I am removing the header from these methods. This is a leftover. I was not able to find a reason to keep them.

File: sources/packages/web/src/services/http/InstitutionUserApi.ts (hunk 11 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/InstitutionUserApi.ts:20]: I am removing the header from these methods. This is a leftover. I was not able to find a reason to keep them.

**Review by ann-aot — COMMENTED** [review]
Good work @andrewsignori-aot  added some comments

File: sources/packages/web/src/services/http/InstitutionUserApi.ts (hunk 12 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/InstitutionUserApi.ts:55]: There is but it is not used. I will remove it.

File: sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts (hunk 1 above)
Comment (ann-aot) [reply sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts:241]: oh..okay... if the endpoint is only for the state, then it's fine

File: sources/packages/api/src/services/institution/institution.service.ts (hunk 4 above)
Comment (ann-aot) [reply sources/packages/api/src/services/institution/institution.service.ts:170]: oh,, okay..got it

File: sources/packages/web/src/services/http/InstitutionUserApi.ts (hunk 12 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/InstitutionUserApi.ts]: Changed!

File: sources/packages/web/src/services/http/InstitutionUserApi.ts (hunk 11 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/InstitutionUserApi.ts]: Changed!

File: sources/packages/web/src/services/http/InstitutionLocation.ts (hunk 10 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/InstitutionLocation.ts]: Changed!

File: sources/packages/web/src/services/InstitutionUserService.ts (hunk 9 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/InstitutionUserService.ts]: Added!

File: sources/packages/web/src/services/InstitutionUserService.ts (hunk 7 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/InstitutionUserService.ts:117]: Added!

File: sources/packages/api/src/services/institution/institution.service.ts (hunk 5 above)
Comment (ann-aot) [reply sources/packages/api/src/services/institution/institution.service.ts:458]: yup, you are correct. ignore it. 😉

File: sources/packages/web/src/services/InstitutionUserService.ts (hunk 6 above)
Comment (ann-aot) [reply sources/packages/web/src/services/InstitutionUserService.ts:61]: mm, got from the `typescript-eslint`. tried this?
image

File: sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts:241]: But I am still keen to remove the address part. If we need it in the future we can add it. I do not recall why we added it.
My best guess is that old UI maybe have some address information in the headers.

File: sources/packages/web/src/services/InstitutionUserService.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/InstitutionUserService.ts:61]: Changed as per recommendation. 
image

**Review by ann-aot — APPROVED** [review]
Thanks for doing the changes 👍 Good Work @andrewsignori-aot

File: sources/packages/api/src/services/institution/institution.service.ts [hunk 13]
```diff
@@ -227,13 +195,28 @@ export class InstitutionService extends RecordDataModelService<Institution> {
...
+      });
+    // Assign the new user with an admin authorization.
+    const userAuthorization = new InstitutionUserAuth();
+    userAuthorization.creator = user;
+    userAuthorization.authType = authorizationType;
+    userAuthorization.institutionUser = institutionUser;
+    // Associates the authorizations.
+    institutionUser.authorizations = [userAuthorization];
+    // Saves institution, user, and institution user altogether.
+    await this.institutionUserRepo.save(institutionUser);
```
Comment (henryoforeh-aot) [inline sources/packages/api/src/services/institution/institution.service.ts:220]: Nothing major, but I'm thinking if this should be wrapped (lines 206 to 219) with a transaction manager.

Comment (henryoforeh-aot) [reply sources/packages/api/src/services/institution/institution.service.ts:220]: Please disregard comment. I was thinking there was another save happening somewhere in between. I will resolve.

**Review by henryoforeh-aot — APPROVED** [review]

File: sources/packages/web/src/services/http/dto/Institution-user.dto.ts [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+   */
+  permissions: UserPermissionAPIInDTO[];
+}
+
+/**
+ * Update an existing user association with an institution
+ * changing the authorizations.
+ */
+export type UpdateInstitutionUserAPIInDTO = Omit<
+  CreateInstitutionUserAPIInDTO,
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/http/dto/Institution-user.dto.ts:67]: The CreateInstitutionUserAPIInDTO does not have userId? did u mean bceidUserId?

File: sources/packages/web/src/views/institution/Login.vue [hunk 15]
```diff
@@ -1,49 +1,74 @@
...
+            with your Business BCeId
+          </Message>
+          <Message severity="error" v-if="showDisabledUserMessage">
+            Disabled User - you don't have access to the system. Please contact
+            Administrator for more informations.
+          </Message>
+          <Message severity="error" v-if="showUnknownUserMessage">
+            The user was validated successfully but is not currently allowed to
+            have access to this application. Please contact the Administrator
+            for more information
```
Comment (guru-aot) [inline sources/packages/web/src/views/institution/Login.vue:67]: Please add a period at the end.

File: sources/packages/web/src/views/institution/Login.vue [hunk 16]
```diff
@@ -1,49 +1,74 @@
...
+            src="@/assets/images/person-standing-with-laptop.svg"
+        /></v-col>
+        <v-col cols="12" class="my-n3">
+          <Message severity="error" v-if="showBasicBCeIDMessage">
+            No such Business account has been found with BCeID. Please login
+            with your Business BCeId
+          </Message>
+          <Message severity="error" v-if="showDisabledUserMessage">
+            Disabled User - you don't have access to the system. Please contact
+            Administrator for more informations.
```
Comment (guru-aot) [inline sources/packages/web/src/views/institution/Login.vue:62]: Please change "Administrator for more informations.", to "Administrator for more information."

File: sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts]: @ann-aot I removed the address part from this method and from the state also - FYI 😉

File: sources/packages/api/src/route-controllers/institution-user/institution-user.institutions.controller.ts [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+
+  /**
+   * Get institution user summary.
+   * @param paginationOptions pagination options.
+   * provided the institution id will be retrieved from the token,
+   * when available.
+   * @returns all filtered institution users.
+   */
+  @Get()
+  async searchUsers(
```
Comment (dheepak-aot) [inline sources/packages/api/src/route-controllers/institution-user/institution-user.institutions.controller.ts:104]: The method name says search users. I assume it is getInstitutionuserSummary(as per comment description)?

File: sources/packages/web/src/services/http/dto/Institution-user.dto.ts (hunk 14 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/dto/Institution-user.dto.ts:67]: Yes, `bceidUserId`. I will change it.

File: sources/packages/api/src/route-controllers/institution-user/institution-user.institutions.controller.ts (hunk 17 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/institution-user/institution-user.institutions.controller.ts:104]: And it does search the users also, right? I will change it.

**Review by guru-aot — APPROVED** [review]
LGTM, thanks for doing the changes. Nice work @

File: sources/packages/web/src/components/common/InstitutionUserSummary.vue [hunk 18]
```diff
@@ -92,10 +91,10 @@
             <span class="text-decoration-underline">Edit</span>
           </v-btn>
           <v-btn
-            v-if="canDisableUser(slotProps.data.userName)"
+            :disabled="slotProps.data.disableRemove"
             @click="updateUserStatus(slotProps.data)"
             variant="text"
-            color="primary"
+            :color="slotProps.data.disableRemove ? 'gray' : 'primary'"
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/common/InstitutionUserSummary.vue:97]: Why are we editing the style of disabled behavior? Is it to match figma? Generally we do not edit the looks of button at it's disabled status. Correct me If I am missing something.

Comment (andrewsignori-aot) [reply sources/packages/web/src/components/common/InstitutionUserSummary.vue:97]: Yes, it is to match Figma. Either it is a Vuetify beta issue or the out-of-box behavior, but for these button types the 'disable' does not cause much UI change and does not match Figma.
image
As you can see this button demands a bit more of the out-of-box to look a bit better. I am not sure if the same look will be used all around the application but if the out-of-box will not be easy to achieve we can create a component to reuse it.

File: sources/packages/web/src/services/AuthService.ts [hunk 19]
```diff
@@ -15,6 +14,7 @@ import {
 import { RENEW_AUTH_TOKEN_TIMER } from "@/constants/system-constants";
 import { StudentService } from "@/services/StudentService";
 import { useStudentStore } from "@/composables";
+import { InstitutionUserService } from "./InstitutionUserService";
```
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/AuthService.ts]: Changed!

File: sources/packages/web/src/components/common/InstitutionUserSummary.vue (hunk 18 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/common/InstitutionUserSummary.vue:97]: Ah.. I get it.
